src/messages/mod.rs

```diff
@@ -170,6 +169,48 @@ impl Message {
     pub fn hash(&self) -> &MessageHash {
         &self.hash
     }
+
+    // Extend the current message proof so it starts at `new_first_key` while keeping the last key
+    // (and therefore the signature) intact.
+    #[cfg_attr(feature = "mock_base", allow(clippy::trivially_copy_pass_by_ref))]
+    pub(crate) fn extend_proof_chain(
+        mut self,
+        new_first_key: &bls::PublicKey,
+        section_proof_chain: &SectionProofChain,
+    ) -> Result<Self, ExtendProofChainError> {
+        let proof_chain = match &mut self.src {
+            SrcAuthority::Section { proof, .. } => proof,
+            SrcAuthority::Node { .. } => return Err(ExtendProofChainError::MustBeSection),
+        };
+
+        if proof_chain.has_key(new_first_key) {
+            return Err(ExtendProofChainError::AlreadySufficient);
+        }
+
+        let index_from = if let Some(index) = section_proof_chain.index_of(new_first_key) {
+            index
+        } else {
+            return Err(ExtendProofChainError::InvalidFirstKey);
+        };
+
+        let index_to = if let Some(index) = section_proof_chain.index_of(proof_chain.last_key()) {
+            index
+        } else {
+            return Err(ExtendProofChainError::InvalidLastKey);
+        };
+
+        *proof_chain = section_proof_chain.slice(index_from..=index_to);
+
+        self.reserialise()
```

Comment:
PErhaps rather than this method, we use the `new_signed` method as it will serialise and hash, etc. Then we are more in line with immutable messages. So we can never change interior code of a msg, but always create a new message?

Reply:
Yeah,I was considering that, but the issue is with error handling. `new_signed` returns `Result<Self, RoutingError>` but I wanted to return a more specific error here. There are ways to have both, but I'm not sure it was worth spending time on it.

Reply:
Can we then match or map this error to allow us to pass a new error? Mind you I suppose that kinda breaks our error handling / types. So the choice is to break error handling or break msg immutability? 
I would perhaps vote that if possible can we keep msg immutability and if required create a new error?

Reply:
What I think I can do is to change `new_signed` to return a separate error, say `CreateError` and define `From` conversions of `CreateError` -> `RoutingError` and `CreateError` -> `ExtendProofChainError`. That should give us everything we need without breaking anything. How does that sound?

Reply:
I think we can actually use the From macro in the derive_more crate to allow that conversion, just to make it easier.

Reply:
Btw, `err-derive` has it's own `From` thing now, so we don't even need `derive_more` anymore.

Reply:
Ah nice one.  I had an issue using it for errors form bls, but that is due to bls using Failure which is not a Error derived form std::Error, but shoudl be cool here.

Reply:
So it seems that normally you can use the `#[error(source)]` attribute which gives you `From` + `std::Error`'s `source`, but if the underlying type doesn't implement `std::Error` you can use `#[error(from)]` which just gives you `From`.

Reply:
Ah, good to know, I wrestled with that for an hour last wekk. Cheers Adam :+1: